Collect unknown keywords as annotations

handrews opened the pull request:
Fixes #24.  This supports collecting unknown keywords as annotation by encapsulating keyword class retrieval into a method, which dynamically creates classes prefixed with `UnknownKeyword` that inherit from `AnnotationKeyword`, and caches the results with the other keyword classes.

To avoid circular dependencies, this  moves the `AnnotationKeyword` class from the `jschon.vocabulary.annotation` module into the `jschon.vocabulary` module with the `Applicator` and `ApplicatorMixin` classes, and splits it similarly into `AnnotationKeyword` and `AnnotationMixin`.

`AnnotationKeyword` is specifically a keyword that is _only_ an annotation, while `AnnotationMixin` stores the annotation but does not preclude further action by subclasses.

Since `AnnotationKeyword` is still imported into `jschon.vocabulary.annotation`, any code relying on being able to import it from there should still work.  There do not appear to be any such imports within jschon or its tests.

There are other ways this could be tested, but the included test covers the changed code, and I'm planning to add more tests around vocabularies and meta-schemas after analyzing the coverage reports in more detail.

------
```
------------------------------------------------------------------------------------------------- benchmark: 5 tests -------------------------------------------------------------------------------------------------
Name (time in us)                      Min                    Max                  Mean                StdDev                Median                 IQR            Outliers          OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_create_json                   40.2911 (1.0)      19,984.2080 (43.29)       56.1091 (1.0)        226.0353 (6.09)        41.5000 (1.0)        2.4170 (1.0)        2;1775  17,822.4145 (1.0)        7921           1
test_evaluate_json[valid]         153.8752 (3.82)     14,723.1249 (31.90)      173.5659 (3.09)       215.8083 (5.81)       159.3751 (3.84)       4.3022 (1.78)       23;732   5,761.5018 (0.32)       4649           1
test_evaluate_json[invalid]       181.5001 (4.50)        461.5840 (1.0)        201.3841 (3.59)        37.1386 (1.0)        188.1660 (4.53)       4.1146 (1.70)      717;750   4,965.6353 (0.28)       4191           1
test_create_schema                481.2079 (11.94)    24,520.5411 (53.12)      545.7938 (9.73)       872.3568 (23.49)      495.7090 (11.94)     23.0831 (9.55)         3;41   1,832.1938 (0.10)       1426           1
test_validate_schema            2,489.9999 (61.80)    41,206.1249 (89.27)    3,315.5801 (59.09)    4,535.8887 (122.13)   2,540.2079 (61.21)    634.0828 (262.34)        5;5     301.6063 (0.02)        327           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
========================== 2399 passed in 106.70s (0:01:46) ==========================
py310: commands_post[0]> coverage report
Name                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------
jschon/__init__.py                   20      2      2      0    91%
jschon/catalog/_2019_09.py           18      0      0      0   100%
jschon/catalog/_2020_12.py           18      0      0      0   100%
jschon/catalog/__init__.py          136     11     30      3    89%
jschon/catalog/_next.py              18      0      0      0   100%
jschon/exceptions.py                 14      0      0      0   100%
jschon/formats.py                     7      0      4      1    91%
jschon/json.py                      210     11     86      4    95%
jschon/jsonpatch.py                 169     43     60     15    70%
jschon/jsonpointer.py               128      6     40      2    95%
jschon/jsonschema.py                263     18    121     11    90%
jschon/output.py                     80      2     41      1    96%
jschon/uri.py                        68      2     14      2    95%
jschon/utils.py                      40      5     12      2    87%
jschon/vocabulary/__init__.py        84      6     24      5    88%
jschon/vocabulary/annotation.py      32      0      2      0   100%
jschon/vocabulary/applicator.py     254      0    147      2    99%
jschon/vocabulary/core.py           123     14     36      8    85%
jschon/vocabulary/format.py          28      0      4      0   100%
jschon/vocabulary/legacy.py         100      2     58      5    96%
jschon/vocabulary/validation.py     159      0     68      1    99%
-------------------------------------------------------------------
TOTAL                              1969    122    749     62    92%
py310: commands_post[1]> coverage html
Wrote HTML report to htmlcov/index.html
.pkg: _exit> python /Users/handrews/dev/.venv/lib/python3.8/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
  py38: OK (117.18=setup[1.05]+cmd[115.55,0.21,0.37] seconds)
  py39: OK (110.26=setup[1.58]+cmd[108.14,0.20,0.33] seconds)
  py310: OK (108.97=setup[1.05]+cmd[107.37,0.20,0.36] seconds)
  congratulations :) (336.51 seconds)
```

marksparkza commented on jschon/vocabulary/__init__.py:
```diff
@@ -51,6 +51,22 @@ def _bootstrap(self, value: Mapping[str, JSONCompatible]) -> None:
             for vocabulary in self.default_vocabularies:
                 self.kwclasses.update(vocabulary.kwclasses)
 
+    def get_kwclass(self, key):
```
Please add type hints for the param and return value.

Could you also please add a short docstring for this method? I know this module is a bit light on docstrings, but ultimately anything that's part of the public API should have one.

handrews replied:
Will do!  I'm trying to be mindful about the type hinting but it's still new to me and sometimes I forget.  Agree about the docstrings.

marksparkza commented on jschon/vocabulary/__init__.py:
```diff
@@ -51,6 +51,22 @@ def _bootstrap(self, value: Mapping[str, JSONCompatible]) -> None:
+                (key[1:] if len(key) > 1 else "")
+            )
+            unknown_class = type(
+                f'UnknownKeyword{capitalized_name}',
```
There's a Python shortcut for this :wink: Check out `str.title()`

handrews replied:
I'm familiar with it, and sadly...
```Python
>>> "unknownKeyword".title()
'Unknownkeyword'
```
😭 
We need a `str.camel()` or something.

marksparkza replied:
Ah, gotcha. In that case the following will suffice:
```py
capitalized_name = key[0].upper() + key[1:]
```
A non-existent slice of a string just gives an empty string, there's no KeyError to avoid.

marksparkza commented on jschon/vocabulary/__init__.py:
```diff
@@ -51,6 +51,22 @@ def _bootstrap(self, value: Mapping[str, JSONCompatible]) -> None:
+            unknown_class = type(
+                f'UnknownKeyword{capitalized_name}',
+                (AnnotationKeyword,),
+                {'key': key},
```
Nice one! I like this approach.

I know this is really pedantic, but I'd prefer `dict(key=key)` here: I think it makes it a bit more obvious that this becomes the `__dict__` of the new class, and that it is setting a class attribute.

handrews replied:
Makes sense to me - it's your project, please be as pedantic as you like!

marksparkza commented on jschon/vocabulary/__init__.py:
```diff
@@ -105,6 +121,17 @@ def __str__(self) -> str:
 KeywordClass = Type[Keyword]
 
 
+class AnnotationMixin(Keyword):
```
I'm going to suggest dropping `AnnotationMixin` and just going with `AnnotationKeyword(Keyword)`, with the same class definition as it had within the annotation.py module. There are two aspects to this:
1. A mixin class is usually not used as a primary ancestor class; it doesn't form part of the class hierarchy of whatever it's being mixed into (and I think trying call a mixin method via `super()` usually doesn't make sense); it adds rather than modifies behaviour.
2. In light of point 1, a better name for `AnnotationMixin` in this context might be something like `AbstractAnnotationKeyword` or `BaseAnnotationKeyword`, but I don't think this level of abstraction is needed (until such time as there is a use case for it).

handrews replied:
I'm happy to do this, although I have some thoughts and questions.  I'm also happy to implement what you've said for this PR and move this discussion out to an issue if you think there's more discussion worth having.

I definitely agree that `AbstractAnnotationKeyword` would be a better name.  It occurred to me that implementing the method made it different from `ApplicatorMixin`** but I didn't think of anything to do about it at the time.

I made `AnnotationMixin` with an eye towards the API for extension keywords, which might both annotate and do something else.  And actually, I forgot that there is already a use case for it: `format`, which (depending on the vocabulary and configuration) can be both an annotation and assertion (I try t forget that `format` exists at all, or I would have refactored `FormatKeyword` to inherit from it.  That would be the correct usage to demonstrate for custom keyword authors, and if you want to keep `AbstractAnnotationKeyword` I would do that refactoring.

**if I may be pedantic 😅 , I'd actually suggest renaming `ApplicatorMixin` and friends `SchemaLocationMixin`, etc.  `$defs` uses this hierarchy but is not an applicator, while `$ref` and `$dynamicRef` which are applicators, do not, which I found confusing.  In draft-next (and ambiguously in earlier drafts), [`contentSchema` ought to inherit from this mixin as well](https://github.com/json-schema-org/json-schema-spec/pull/1352).  If you'd be receptive to such a change, I can file a separate issue and make a PR.

marksparkza replied:
Could you please create an issue for renaming the applicator mixin classes? I agree in principle, though I'm thinking of renaming as follows (but I'm open to discussion on this):

ApplicatorMixin -> SubschemaMixin
Applicator -> Subschema
ArrayApplicator -> ArrayOfSubschemas
PropertyApplicator -> ObjectOfSubschemas

Then, class definitions would look like:

```py
class IfKeyword(Keyword, Subschema):
    ...

class AnyOfKeyword(Keyword, ArrayOfSubschemas):
    ...

class PropertiesKeyword(Keyword, ObjectOfSubschemas):
    ...
```

And please create a separate issue and PR for the `contentSchema` change.

Thanks!

handrews replied:
Will do - this is also related to the stuff I alluded to regarding `"propertyDependencies"` (which is an object of object of schemas instead of just an object of schemas), so I might take a day or two while I make sure I'm covering everything.  If you can take a look at PRs 57, 65, 66, and 67 (which are all quite short and simple) in the meantime that will simplify my life immensely due to git's awful submodule behavior (why do submodules not automatically update to match the main repo's branch state?)

marksparkza replied:
>  If you can take a look at PRs 57, 65, 66, and 67

Certainly - I should be able to get through them tomorrow.

> git's awful submodule behavior

I don't understand it either :shrug:

marksparkza replied:
OK, I've merged 57. I'd like to spend a bit more time on 65 and 66, however, to fully understand the issues they resolve and the implications of the changes. I'll only have time for this on the weekend, unfortunately.

handrews replied:
@marksparkza no problem!  Both of those PRs address JSON Schema Test Suite failures - I'll update the descriptions with the details.

marksparkza commented on jschon/vocabulary/__init__.py:
```diff
@@ -105,6 +121,17 @@ def __str__(self) -> str:
+        result.annotate(self.json.data)
+
+
+class AnnotationKeyword(AnnotationMixin):
```
Please add `'AnnotationKeyword'` to the list of exported module members in `__all__`.

handrews replied:
Oop!  I forget about `__all__` as I've never used it much (in the last really big Python project I worked on we banned using `import *` instead).

handrews commented:
@marksparkza I think I have addressed all of your concerns with this latest commit, including removing `AnnotationMixin` (if you find my argument for `AbstractAnnotationKeyword` compelling I can add it back in, or do it as a separate PR).

marksparkza commented on jschon/vocabulary/annotation.py:
```diff
@@ -3,6 +3,7 @@
 from jschon.vocabulary import AnnotationKeyword
 
 __all__ = [
+    'AnnotationKeyword',
```
Sorry, I meant add it to `__all__` in `vocabulary/__init__.py`. It should be exported from the module where it's defined.

`__all__` is not only relevant to `import *`, perhaps more importantly it represents the public API of a module. I picked this up because my IDE highlights imports of non-public module members:

![__all__](https://user-images.githubusercontent.com/52427991/223030644-726cf580-e241-4bf9-9041-0d6797ab9c4e.png)

marksparkza commented:
I'm having second thoughts about re-using `AnnotationKeyword` outside of the `annotation` vocab module. It was only intended to be internal to that module to eliminate evaluation code repetition. It was intentionally left out of `__all__` and should perhaps have been named `_AnnotationKeyword`. The `Keyword` class hierarchy is otherwise flat and I think it's better to keep it that way. Aside from the framework integration that the `Keyword` class itself provides (with the help of mixins where applicable), the `Result` API makes it easy for each keyword (including any custom/extension keyword) to declare its own behaviour entirely and explicitly.

I don't think there's a strong case for partitioning the `Keyword` class hierarchy into subordinate abstract classes. In my view, there's no clear distinction between annotation and non-annotation keywords, for example. All keywords are capable of creating annotations, or capable (in principle) of being extended to create annotations.

With respect to unknown keywords, my suggestion would be an `UnknownKeyword` (or perhaps `_UnknownKeyword`) base class with the appropriate annotation behaviour, and private to `vocabulary.__init__`, from which the concrete dynamic unknown keywords can be derived.

Does this make sense?

handrews commented:
@marksparkza I follow your reasoning and I'm happy to adjust this PR to match it.

I do think there is a strong argument for additional base classes in the hierarchy to facilitate extension vocabularies (which, TBH, is my primary reason for using and contributing to `jschon` - it's a well-designed Python3 library and AFAIK the only Python implementation with a functioning vocabulary system).

But I think it's worth separating that discussion from this more focused PR, so I'll update this as you request (also, while I'll always propose ideas if I think they would aid the project, I'm not trying to take over the project so even if I think something would really be better, I'm happy to drop it if it does not fit your vision for jschon).

handrews commented:
@marksparkza updated with _UnknownKeyword, which I assume isn't supposed to have a docstring as an internal class (lmk if you prefer otherwise, of course!).

The force push is just because I rebased things (no conflicts).  Only the new commit has new changes, the others are identical to what they were before.

marksparkza commented:
Thanks!

marksparkza commented:
> I do think there is a strong argument for additional base classes in the hierarchy to facilitate extension vocabularies 

The question of whether or not to deepen the keyword class hierarchy is an interesting one, and it got me thinking more carefully about why I'm inclined towards keeping it flat.

> your vision for jschon

In developing jschon, I’ve drawn a lot of inspiration from two great Python libraries that I use extensively in my day job: Pydantic and SQLAlchemy.

Applications that use Pydantic or the ORM paradigm in SQLAlchemy are easy to spot: you’ll find a collection of declarative models, each derived from a quintessential base class provided by the library - `BaseModel` in Pydantic, `DeclarativeBase` in SQLAlchemy. A given application might define a model hierarchy to roll up common behaviours, but that would be private to the application and, in general, should not affect the relationships between the declarative models and other foundational entities in the library. One of the great things about implementing declarative models using these libraries is that I know that I can just derive from `BaseModel` or `DeclarativeBase`, I don't need to know which is the most appropriate abstract subclass to use, and I don't need to be aware of any variance in implementation higher up in the hierarchy. Another great thing is that each model is completely self-contained and completely self-descriptive.

In jschon, keywords are analogous to declarative models, with `Keyword` being the quintessential base class. I realise that you could argue that declarative models are conceptually more analogous to schemas, and model fields to keywords, but I'm looking at things from a programming workflow perspective: as declarative models are the building blocks of a Pydantic / SQLAlchemy application, so keywords are the building blocks of a vocabulary implementation. Taking the analogy further, a given vocabulary might define a keyword class hierarchy to roll up common behaviours, but that should be private to the module, extension or application that implements the vocabulary, and should not affect the underlying relationships between keywords and other foundational entities in the jschon library. From the perspective of those other entities, a keyword is a keyword is a keyword, and from the perspective of the vocabulary, a keyword is self-contained and self-descriptive.

handrews commented:
@marksparkza thanks for explaining your philosophy for the library — I find that very helpful for understanding how to approach changes and what ideas might fit best.  (It might be worth adding some of that to CONTRIBUTING.md).

Yesterday I started writing up an issue to file about the hierarchy and, if anything, started feeling less strongly about it.  I ended up stuffing the half-written idea in a file to reconsider later instead of submitting the issue.

I think I will eventually submit one issue that has to do with difficulties I'm experiencing implementing `propertyDependencies` for draft-next.  But that's less about the hierarchy and more about how many places I needed to touch to get an applicator with a new subschema arrangement to work.  I'm pretty sold on _not_ doing an `AbstractAnnotationKeyword` or similar class.
